lib/Onnxifi/Base.cpp

```diff
@@ -43,6 +47,53 @@ void Event::wait() {
   cond_.wait(guard, [this] { return fired_ == true; });
 }
 
+constexpr std::chrono::microseconds ThreadPool::kThreadBreakDuration;
+
+ThreadPool::ThreadPool(unsigned numWorkers) : shouldStop_(false) {
+  // Intialize all workers and make each one run threadPoolWorkerMain.
+  for (unsigned i = 0; i < numWorkers; i++) {
+    std::thread th(std::bind(&ThreadPool::threadPoolWorkerMain, this));
+    workers_.push_back(std::move(th));
+  }
+}
+
+ThreadPool::~ThreadPool() {
+  // Signal to workers to stop.
+  shouldStop_ = true;
+
+  // Join all worker threads.
+  for (auto &w : workers_) {
+    w.join();
+  }
+  workers_.clear();
+}
+
+void ThreadPool::submit(const std::function<void(void)> &fn) {
+  // Add fn to the work queue.
+  std::lock_guard<std::mutex> lock(workQueueMtx_);
+  workQueue_.push(std::move(fn));
+}
+
+void ThreadPool::threadPoolWorkerMain() {
+  std::unique_lock<std::mutex> lock(workQueueMtx_, std::defer_lock);
+
+  while (!shouldStop_) {
+    lock.lock();
+
+    if (!workQueue_.empty()) {
+      // Pop and process a work item.
+      auto workItem = std::move(workQueue_.front());
+      workQueue_.pop();
+      lock.unlock();
+      workItem();
```

Comment:
So, we're currently compiling glow with `-fno-exceptions`.  I'm not sure this is a good idea, since we're likely to be embedded in systems that *do* use exceptions.

Reply:
I just realized what I wrote was ambiguous.  To clarify, I meant, I'm not sure compiling with `-fno-exceptions` is a good idea long-term.  It probably is a good idea to try/catch the `workItem()` call.

Reply:
Hmm, actually it this is the case, it means that the workItem should will not throw whatsoever because itself is built with `-fno-exceptions`, so we should be good without `try/catch`.